src/cpp/include/openvino/genai/parsers.hpp

```diff
@@ -0,0 +1,96 @@
+// Copyright (C) 2023-2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#include <string>
+#include <memory>
+#include <variant>
+#include <map>
+#include <optional>
+#include <vector>
+
+namespace ov {
+namespace genai {
+
+
+using ParsedMessage = std::map<std::string, std::string>;
+
+class IncrementalParserBase {
+public:
+    IncrementalParserBase() = default;
+
+    virtual ParsedMessage parse(
+        ParsedMessage& msg,
+        const std::string& previous_text, 
+        const std::string& delta_text, 
+        const std::optional<std::vector<int64_t>>& previous_tokens = std::nullopt, 
+        const std::optional<std::vector<int64_t>>& delta_tokens = std::nullopt
+    ) = 0;
+
+    virtual bool is_active() const = 0;
+    static std::map<std::string, std::shared_ptr<IncrementalParserBase>> registered_parsers;
+};
+
+class DeepSeekR1ReasoningParser : public IncrementalParserBase {
+private:
+    bool m_starts_with_thinking = true;
+    bool m_think_tag_opened = false;
+    bool m_deactivated = false;
+    std::string m_open_tag = "<think>";
+    std::string m_close_tag = "</think>";
+public:
+    DeepSeekR1ReasoningParser(bool starts_with_thinking = true) : m_starts_with_thinking(starts_with_thinking) {};
+    std::map<std::string, std::string> accumulated_parsed;
+
+    ParsedMessage parse(
+        ParsedMessage& msg,
+        const std::string& previous_text, 
+        const std::string& delta_text,
+        const std::optional<std::vector<int64_t>>& previous_tokens = std::nullopt, 
+        const std::optional<std::vector<int64_t>>& delta_tokens = std::nullopt
+    ) override;
+    static std::string name() { return "DeepSeekR1ReasoningParser"; }
+    bool is_active() const override;
+};
+
+class ParserBase {
+public:
+    ParserBase() = default;
+
+    virtual ParsedMessage parse(ParsedMessage& text) = 0;
+    static std::map<std::string, std::shared_ptr<ParserBase>> registered_parsers;
```

Comment:
`registered_parsers` is a global variable eesentially. Modifying it affects the available pipelines. I think if someone wanted to remove a parser, we would encourage them to do that for their instance of pipeline via GeneratioConfig. As for extending with new parsers, it can be solved with a function.

`generate()` uses the parsers from `registered_parsers` but parsers may have a state. If there are two `LLMPipeline`s, this can be a problem